build/i2c/src/lib.rs

```diff
@@ -1358,6 +1404,115 @@ impl ConfigGenerator {
         Ok(())
     }
 
+    fn declare_sensor_struct(
+        &mut self,
+        d: &I2cDevice,
+        struct_name: &str,
+    ) -> Result<()> {
+        // Manually unpack the field so that changes to the sensor types
+        // will require changes here as well.
+        if let Some(I2cSensors {
+            temperature,
+            power,
+            current,
+            voltage,
+            input_current,
+            input_voltage,
+            speed,
+            names: _,
+        }) = &d.sensors
+        {
+            writeln!(
+                &mut self.output,
+                "\n        #[allow(non_camel_case_types)]
+        pub struct Sensors_{struct_name} {{",
+            )?;
+            let mut f = |name, count| match count {
+                0 => Ok(()),
+                1 => writeln!(
+                    &mut self.output,
+                    "            pub {name}: SensorId,"
+                ),
+                _ => writeln!(
+                    &mut self.output,
+                    "            pub {name}: [SensorId; {count}],"
+                ),
+            };
+            f("temperature", *temperature)?;
+            f("power", *power)?;
+            f("current", *current)?;
+            f("voltage", *voltage)?;
+            f("input_current", *input_current)?;
+            f("input_voltage", *input_voltage)?;
+            f("speed", *speed)?;
+            writeln!(&mut self.output, "        }}")?;
+        } else {
+            writeln!(
+                &mut self.output,
+                "\n        #[allow(dead_code, non_camel_case_types)]
+        type Sensors_{struct_name} = ();",
+            )?;
+        }
+        Ok(())
+    }
+
+    fn emit_sensor_struct(
+        &mut self,
+        d: &I2cDevice,
+        label: String,
+        name: &str,
+        sensors: &[DeviceSensor],
+    ) -> Result<()> {
+        write!(
+            &mut self.output,
+            "        #[allow(dead_code)]
+        pub const {}_{label}_SENSORS: Sensors_{name} = ",
+            d.device.to_uppercase(),
+        )?;
+
+        let mut sensors_by_kind: HashMap<Sensor, Vec<usize>> = HashMap::new();
```

Comment:
Since we iterate over this below, maybe a `BTreeMap` so the order is consistent between runs? It doesn't look like that matters for correctness, but might be nicer for humans skimming the output.

Reply:
Good idea, done in ea841d070f085bfb8a68d362659c54d386e0253a